lib/handleRequest.js

```diff
@@ -11,140 +11,118 @@ const inputSchemaError = fastJsonStringify(schemas.inputSchemaError)
 
 function handleRequest (req, res, params, context) {
   var method = req.method
+  var request = new context.Request(params, req, urlUtil.parse(req.url, true).query, req.headers, req.log)
+  var reply = new context.Reply(res, context, request)
 
   if (method === 'GET' || method === 'HEAD') {
-    return handler(context, params, req, res, null, urlUtil.parse(req.url, true).query)
+    return handler(reply)
   }
 
+  var contentType = req.headers['content-type']
+
   if (method === 'POST' || method === 'PUT' || method === 'PATCH') {
     // application/json content type
-    if (req.headers['content-type'] && req.headers['content-type'].indexOf('application/json') > -1) {
-      return jsonBody(req, res, params, context)
+    if (contentType && contentType.indexOf('application/json') > -1) {
+      return jsonBody(request, reply)
     }
 
     // custom parser for a given content type
-    if (context.contentTypeParser.fastHasHeader(req.headers['content-type'])) {
-      return context.contentTypeParser.run(req.headers['content-type'], handler, context, params, req, res, urlUtil.parse(req.url, true).query)
+    if (context.contentTypeParser.fastHasHeader(contentType)) {
+      return context.contentTypeParser.run(contentType, handler, request, reply)
     }
 
-    wrapReplyEnd(context, req, res, 415, params, null, urlUtil.parse(req.url, true).query, null)
+    reply.code(415).send(new Error('Unsupported Media Type: ' + contentType))
```

Comment:
this is where we could use http-errors @jsumners

Reply:
I don't think we should use `http-errors` directly in core, is a useful utility for the users, but here we know what error we want to send as well as the status code.
Use our core API is definitely faster than use an external error generator.

Reply:
And there are so few instances where core generates an error that it just isn't worth the added dependency (and the added code paths).